[choreolib, docs] ChoreoLib docs overhaul (#994)
SleipnirGroup/Choreo: docs/choreolib/auto-factory.md

```java
// Follows deploy/choreo/myTrajectory.traj
Command myTrajectory = autoFactory.trajectoryCommand("myTrajectory");
```

!!! warning
    Trajectories are sampled using a timer, which also dictates when the command ends. This means that your robot is technically not guaranteed to finish exactly at the end of a trajectory. Properly tuning your robot's drivetrain as well as feedback gains in your trajectory follower will mitigate this error.

You can then combine trajectory commands with other functions of your robot to create an autonomous routine. For more information about command compositions, check out [WPILib's documentation](https://docs.wpilib.org/en/stable/docs/software/commandbased/command-compositions.html).

```java
public Command pickupAndScoreAuto() {
    return Commands.sequence(
        // TODO Reset Odometry (Pending API change)
        Commands.deadline(
            autoFactory.trajectoryCommand("pickupGamepiece"),
            intakeSubsystem.intake() // (1)
        ),
        Commands.parallel(
            autoFactory.trajectoryCommand("scoreGamepiece"),
            scoringSubsystem.getReady()
        )
        scoringSubsystem.score()
    );
}
```

> 1. Throughout this documentation, we assume you are using [command factory methods](https://docs.wpilib.org/en/stable/docs/software/commandbased/organizing-command-based.html#instance-command-factory-methods) in your subsystems. For the uninitiated, this is WPILib's recommended alternative to [command subclasses](https://docs.wpilib.org/en/stable/docs/software/commandbased/organizing-command-based.html#writing-command-classes) (i.e. `new IntakeCommand()`).

Reviewer: ```suggestion
1. Throughout this documentation, we assume you are using [command factory methods](https://docs.wpilib.org/en/stable/docs/software/commandbased/organizing-command-based.html#instance-command-factory-methods) in your subsystems. This is a more concise way of describing command based than command classes.
```

## Using AutoRoutine

> While command compositions may be an effective architecture in most cases, it has a tendency to become unwieldy if your autonomous routine has branches, if your subsystems use default commands, or if you want your subsystems to run independently and "hand off" actions to each other. The `AutoRoutine` ([Java](/api/choreolib/java/choreo/auto/AutoRoutine.html)) class aims to solve these problems, using WPILib's [Trigger](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/button/Trigger.html) to define a control flow based on reactions to state that don't require subsystems until they are needed.

Reviewer: ```suggestion
While command compositions may be an effective architecture in some cases, it has a tendency to become unwieldy if your autonomous routine has branches, if your subsystems use default commands, or if you want your subsystems to run independently and "hand off" actions to each other. The `AutoRoutine` ([Java](/api/choreolib/java/choreo/auto/AutoRoutine.html)) class aims to solve these problems, using WPILib's [Trigger](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/button/Trigger.html) to define a control flow based on reactions to state that don't require subsystems until they are needed.
```

or

```suggestion
While command compositions may be an effective architecture in many cases, it has a tendency to become unwieldy if your autonomous routine has branches, if your subsystems use default commands, or if you want your subsystems to run independently and "hand off" actions to each other. The `AutoRoutine` ([Java](/api/choreolib/java/choreo/auto/AutoRoutine.html)) class aims to solve these problems, using WPILib's [Trigger](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/button/Trigger.html) to define a control flow based on reactions to state that don't require subsystems until they are needed.
```

```java
public AutoRoutine pickupAndScoreAuto() {
    AutoRoutine routine = autoFactory.newRoutine("pickupAndScore");

    // Load the routine's trajectories
    AutoTrajectory pickupTraj = routine.trajectory("pickupGamepiece");
    AutoTrajectory scoreTraj = routine.trajectory("scoreGamepiece");

    // When the routine begins, start the first trajectory
    routine.running().onTrue(pickupTraj.cmd()); // TODO Reset Odometry (Pending API change)

    // Starting at the event marker named "intake", run the intake
    pickupTraj.atPose("intake").onTrue(intakeSubsystem.intake());

    // When the trajectory is done, start the next trajectory
    pickupTraj.done().onTrue(scoreTraj.cmd());

    // While the trajectory is active, prepare the scoring subsystem
    scoreTraj.active().whileTrue(scoringSubsystem.getReady());

    // When the trajectory is done, score
    scoreTraj.done().onTrue(scoringSubsystem.score());

    return routine;
}
```

> Sometimes, you may want to implement a "branching auto": an autonomous routine that changes behavior based on the state of the robot. An excellent example for the need of branching autos is the [2024 season](https://youtu.be/9keeDyFxzY4), where robots would race to the midline to grab a gamepiece. If another robot beat yours to the midline, or your robot missed a game piece, a common strategy was to go directly to the next gamepiece on the midline, instead of coming back to score.

Reviewer: A video of the actual autos your describing would be better here

Author: Funnily enough when I originally wrote this section I spent ~20 minutes trying to find an example from Chezy Champs (for better video quality) but unfortunately I couldn't find any good examples, I'll take a look at MTTD to see if I can find anything there, if you have a match in mind let me know and I'll add it in, this was another task I forgot to add to the to do list

Reviewer: I don't have a match but I remember a YouTube short which probably describes it well enough if you can't find a match

    // Run the intake when we are approaching a gamepiece
    C2toM1.active()
        .or(scoreToM2.active())
        .or(scoreToM3.active())
        .or(M1toM2.active())
        .or(M2toM3.active())
        .whileTrue(intakeSubsystem.intake());

    // If we picked up the gamepiece, go to the "Score" location
    // If we didn't pick up the gamepiece, go to the next midline location
    C2toM1.done().onTrue(either(M1toScore.cmd(), M1toM2.cmd(), shooterSubsystem::hasGamepiece));
    scoreToM2.done().or(M1toM2.done()).onTrue(either(M2toScore.cmd(), M2toM3.cmd(), shooterSubsystem::hasGamepiece));
    scoreToM3.done().or(M2toM3.done()).onTrue(either(M3toScore.cmd(), Commands.none(), shooterSubsystem::hasGamepiece));

    // After we go to score, shoot then go to the next available midline location
    M1toScore.done().onTrue(shooterSubsystem.shoot().andThen(scoreToM2.cmd()));
    M2toScore.done().onTrue(shooterSubsystem.shoot().andThen(scoreToM3.cmd()));
    M3toScore.done().onTrue(shooterSubsystem.shoot());

    return routine;
}
```

1. ![Samples view layer showing ](../media/choreolib-branching-auto.png)

## AutoChooser

> The `AutoChooser` ([Java](/api/choreolib/java/choreo/auto/AutoChooser.html)) class allows you to send a list of your autonomous routines to a driver dashboard for selection before a match. It is meant to be a more efficient alternative to `SendableChooser`, taking a [lazy loading](https://en.wikipedia.org/wiki/Lazy_loading) approach to generating command compositions or an `AutoRoutine`. This approach has the benefit of not loading all autos on startup, but also not loading the auto after the match starts, which may cause a delay when using many or large trajectories.

Reviewer: Note this is the recommended pattern for users.

> The `AutoChooser` ([Java](/api/choreolib/java/choreo/auto/AutoChooser.html)) class allows you to send a list of your autonomous routines to a driver dashboard for selection before a match. It is meant to be a more efficient alternative to `SendableChooser`, taking a [lazy loading](https://en.wikipedia.org/wiki/Lazy_loading) approach to generating command compositions or an `AutoRoutine`. This approach has the benefit of not loading all autos on startup, but also not loading the auto after the match starts, which may cause a delay when using many or large trajectories.

Reviewer: ```suggestion
The `AutoChooser` ([Java](/api/choreolib/java/choreo/auto/AutoChooser.html)) class allows you to send a list of your autonomous routines to a driver dashboard for selection before a match. It is meant to be a more efficient alternative to `SendableChooser`, taking a [lazy loading](https://en.wikipedia.org/wiki/Lazy_loading) approach to generating command compositions or an `AutoRoutine`. This approach has the benefit of not loading all autos on startup and not causing your robot to go into a constant crash loop if you have a typo in an auto, but also not loading the auto after the match starts, which may cause a delay when using many or large trajectories.
```

Reviewer: It doesn't prevent crashes it prevents crashes that stop your code until redeploy. If it loads on init then the robot will crash and you can't ever recover from that state. AutoChooser prevents that as when the robot restarts you just pick a different auto

Author: That makes sense, I'll try to fit it in but if I can't word it in a nice way I may just omit it, as imho it isn't really an important feature of the API. Boot looping doesn't really present a huge problem in FRC; just pause the console, fix your code, and redeploy.

```java
// TODO Pending https://github.com/SleipnirGroup/Choreo/pull/949
```

## AutoBindings

> The `AutoBindings` ([Java](/api/choreolib/java/choreo/auto/AutoFactory.AutoBindings.html)) class is used to bind event markers in trajectories made by the `AutoFactory` to commands. This is useful if you have simpler actions that you want to trigger at specific points in a trajectory without much thought.

Reviewer: Note these persist across all autos.

    AutoTrajectory M2toScore = routine.trajectory("M2toScore");
    AutoTrajectory M3toScore = routine.trajectory("M3toScore");
    AutoTrajectory scoreToM2 = routine.trajectory("scoreToM2");
    AutoTrajectory scoreToM3 = routine.trajectory("scoreToM3");

    // When the routine starts, reset odometry, shoot the first gamepiece, then go to the "C2" location
    routine.running().onTrue(
        Commands.sequence(
            startToC2.resetOdometry(),
            shooterSubsystem.shoot(),
            startToC2.cmd()
        )
    );

    // Pick up and shoot the gamepiece at the "C2" location, then go to the "M1" location
    startToC2.active().whileTrue(intakeSubsystem.intake());
    startToC2.done().onTrue(shooterSubsystem.shoot().andThen(C2toM1.cmd()));

    // Run the intake when we are approaching a gamepiece
    C2toM1.active()
        .or(scoreToM2.active())
        .or(scoreToM3.active())
        .or(M1toM2.active())
        .or(M2toM3.active())
        .whileTrue(intakeSubsystem.intake());

    // If we picked up the gamepiece, go to the "Score" location
    // If we didn't pick up the gamepiece, go to the next midline location

>     C2toM1.done().onTrue(either(M1toScore.cmd(), M1toM2.cmd(), shooterSubsystem::hasGamepiece));

Reviewer 2: maybe not here but somewhere note that either requires the subsystems of both branches, if you want each branch to be completely separate you can do

```java
C2toM1.done().and(shooterSubsystem::hasGamepiece).onTrue(M1toScore.cmd());
C2toM1.done().and(shooterSubsystem::noGamepiece).onTrue(M1toM2.cmd());
```
or smthn, this isn't the best example of where it might matter but you get the idea

Author: I'll replace with the modified code and add a note dialog (blue box) about the requirements nuance

```java
routine.running().onTrue(Commands.print("Started the routine!"));
```

Traditional triggers can also be used in conjunction with an auto routine. However, they will experience "hygiene" issues if not used correctly, causing them to be active even when the routine is not. You must either:

- Use `AutoRoutine.observe()` to ensure the trigger is only active when the routine is running
- Conjoin a trigger to one created by `AutoRoutine` or `AutoTrajectory`

```java
Trigger myTrigger = new Trigger(() -> condition);

// Safe
routine.observe(myTrigger).onTrue(Commands.print("Foo"));
routine.running().and(myTrigger).onTrue(Commands.print("Bar"));

// Unsafe
myTrigger.onTrue(Commands.print("Foo"));
myTrigger.and(routine.running()).onTrue(Commands.print("Bar"));
```

Trajectories can be loaded using `AutoRoutine.trajectory()`, which will return an `AutoTrajectory` ([Java](/api/choreolib/java/choreo/auto/AutoTrajectory.html)). The `AutoTrajectory` class exposes multiple triggers for you to attach reactive logic to, as well as `AutoTrajectory.cmd()` for scheduling the trajectory.

```java

> public AutoRoutine pickupAndScoreAuto() {

Reviewer: Can you show some `atTime` examples as well?

Author: I meant to use `atTime()` originally in the example, I made the change and added an annotation for `atPose()`

    AutoRoutine routine = autoFactory.newRoutine("pickupAndScore");

    // Load the routine's trajectories
    AutoTrajectory pickupTraj = routine.trajectory("pickupGamepiece");
    AutoTrajectory scoreTraj = routine.trajectory("scoreGamepiece");

    // When the routine begins, reset odometry and start the first trajectory (1)
    routine.running().onTrue(pickupTraj.resetOdometry().andThen(pickupTraj.cmd()));

    // Starting at the event marker named "intake", run the intake
    pickupTraj.atPose("intake").onTrue(intakeSubsystem.intake());

    // When the trajectory is done, start the next trajectory
    pickupTraj.done().onTrue(scoreTraj.cmd());

    // While the trajectory is active, prepare the scoring subsystem
    scoreTraj.active().whileTrue(scoringSubsystem.getReady());

    // When the trajectory is done, score
    scoreTraj.done().onTrue(scoringSubsystem.score());

    return routine;
}
```

> 1. You should always reset your robot's odometry to the start of the first trajectory being followed in an autonomous routine. `AutoTrajectory.resetOdometry()` will accomplish this, setting the robot's pose to the start of the trajectory.

Reviewer: Is this true? some teams might have vision good enough they don't need to
I do like this as general tip though

Author: Teams should only be seeding their field position with vision if they really know what they're doing, and as far as I know the number of teams that actually did this in 2024 could probably be counted on my fingers. We also don't know what AprilTag layouts will look like in future seasons, and how optimized they will be for pose estimation from wherever robots start on the field. I could modify the wording to mention vision but idk how much value it adds, if you think it's important I can make the change

    AutoTrajectory M3toScore = routine.trajectory("M3toScore");
    AutoTrajectory scoreToM2 = routine.trajectory("scoreToM2");
    AutoTrajectory scoreToM3 = routine.trajectory("scoreToM3");

    // When the routine starts, reset odometry, shoot the first gamepiece, then go to the "C2" location
    routine.active().onTrue(
        Commands.sequence(
            routine.resetOdometry(startToC2),
            shooterSubsystem.shoot(),
            startToC2.cmd()
        )
    );

    // Pick up and shoot the gamepiece at the "C2" location, then go to the "M1" location
    startToC2.active().whileTrue(intakeSubsystem.intake());
    startToC2.done().onTrue(shooterSubsystem.shoot().andThen(C2toM1.cmd()));

    // Run the intake when we are approaching a gamepiece
    C2toM1.active()
        .or(scoreToM2.active())
        .or(scoreToM3.active())
        .or(M1toM2.active())
        .or(M2toM3.active())
        .whileTrue(intakeSubsystem.intake());

    // If we picked up the gamepiece, go score, then go to the next midline location
    // If we didn't pick up the gamepiece, go directly to the next midline location

    // M1

>     Trigger atM1 = C2toM1.done();

Reviewer 3: The behavior that AutoTrajectory.done() is only true for one cycle isn't mentioned anywhere on this page and is key to this example working as desired. This behavior should be explicitly mentioned on this page, as it's critical for users to know.

# AutoFactory (Java Only)

!!! note
    This API is only available for Java. C++ and Python users should instead utilize the [Trajectory API](./trajectory-api.md).

ChoreoLib provides the `AutoFactory` ([Java](/api/choreolib/java/choreo/auto/AutoFactory.html)) class as a higher level API to make it easier to create complex auto routines inside your robot code.

```java title="Robot.java"
public class Robot extends TimedRobot {
    private final Drive driveSubsystem = new Drive();
    private final AutoFactory autoFactory;

    public Robot() {
        autoFactory = new AutoFactory(
            driveSubsystem::getPose, // A function that returns the current robot pose
            driveSubsystem::resetOdometry, // A function that resets the current robot pose to the provided Pose2d
            driveSubsystem::followTrajectory, // The drive subsystem trajectory follower (1)

>             () -> true, // If alliance flipping should be enabled (2)

Reviewer 3: With #1025 changing this to a boolean instead of a supplier, this should be updated to match

    // Load the routine's trajectories
    AutoTrajectory startToC2 = routine.trajectory("startToC2");
    AutoTrajectory C2toM1 = routine.trajectory("C2toM1");
    AutoTrajectory M1toM2 = routine.trajectory("M1toM2");
    AutoTrajectory M2toM3 = routine.trajectory("M2toM3");
    AutoTrajectory M1toScore = routine.trajectory("M1toScore");
    AutoTrajectory M2toScore = routine.trajectory("M2toScore");
    AutoTrajectory M3toScore = routine.trajectory("M3toScore");
    AutoTrajectory scoreToM2 = routine.trajectory("scoreToM2");
    AutoTrajectory scoreToM3 = routine.trajectory("scoreToM3");

    // When the routine starts, reset odometry, shoot the first gamepiece, then go to the "C2" location
    routine.active().onTrue(
        Commands.sequence(
            routine.resetOdometry(startToC2),
            shooterSubsystem.shoot(),
            startToC2.cmd()
        )
    );

    // Pick up and shoot the gamepiece at the "C2" location, then go to the "M1" location
    startToC2.active().whileTrue(intakeSubsystem.intake());
    startToC2.done().onTrue(shooterSubsystem.shoot().andThen(C2toM1.cmd()));

    // Run the intake when we are approaching a gamepiece

>     C2toM1.active()

Reviewer 2: For stuff like this once #1002 is merged, doing a `routine.anyActive(scoreToM2, ...)` should be done to show off the functionality.
